Added cluster setup using go processes

arindamnayak opened the pull request:
Added cluster setup using go processes (vttablet, vtctld, vtgate etc.. processes). 
Added clustertest to verify if cluster setup is working.

arindamnayak committed:
added test for join queries

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Revert "added test for join queries"

This reverts commit f373359689b50d744695a13f3ceb0083ead7e995.

arindamnayak committed:
Added testcases for join

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added comments to code

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

deepthi committed:
Merge pull request #4 from planetscale/tal_vtgate_join

Ported testcases for join queries in vtagatev3_test.py to go

arindamnayak committed:
Initial commit for adding etcd during cluster setup

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added vtctld process

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Added vtctld process during cluster up

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added vttablet process

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
moved port to main configuration

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
updated vtgate and fixed vttablet

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
fixed missed commits

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
added clustersetup as configurable

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
moved cluster code to separate pkg

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
renamed methods

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
updated field name

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

deepthi commented on dev.env:
```diff
@@ -73,6 +73,17 @@ PATH=$(prepend_path "$PATH" "$VTROOT/dist/chromedriver")
 PATH=$(prepend_path "$PATH" "$VTROOT/dist/node/bin")
 export PATH
 
+# Etcd path.
```
Why is this needed?

arindamnayak replied:
Currently `dev.env` does not set path for binary directory of `etcd`. That is why I added here to set the `etcd` path, so that go process can directly spawn `etcd` process.

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -0,0 +1,265 @@
+	nextPortForProcess int
+}
+
+// Keyspace Cluster accepts keyspace to launch it
```
add a `:` after `Keyspace`

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -0,0 +1,265 @@
+	Shards    []Shard
+}
+
+// Shard Keyspace need Shard to manage its lifecycle
```
`Shard with associated vttablets`

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -0,0 +1,265 @@
+	Vttablets []Vttablet
+}
+
+// Vttablet Shard need vttablet to manage the setup and teardown
```
`Vttablet stores the properties needed to start a vttablet process`

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -0,0 +1,265 @@
+		return
+	}
+
+	println("Creating topop dirs")
```
typo: topop -> topo

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -0,0 +1,265 @@
+// Keyspace Cluster accepts keyspace to launch it
+type Keyspace struct {
+	Name      string
+	SQLSchema string
```
This field should be called `SchemaSQL`  instead.

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -0,0 +1,265 @@
+			shard.Vttablets = append(shard.Vttablets, *tablet)
+		}
+
+		// Make 1st tablet as master shard
```
`Make first tablet as master`

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -0,0 +1,265 @@
+			return
+		}
+
+		// Apply SQLSchema
```
`Apply schema sql`

deepthi commented on go/test/endtoend/cluster/etcd_process.go:
```diff
@@ -0,0 +1,160 @@
+}
+
+// ManageTopoDir creates global and zone in etcd2
+func (etcd *EtcdProcess) ManageTopoDir(command string, directory string) error {
```
`command` is not being used here, instead it is always `mkdir`. If the only use of this function is to create directories, then it should be renamed to `CreateTopoDir`, and the fist argument (`command`) can be dropped.

arindamnayak replied:
During `TearDown`, we need to clean up the TopoDir, there the command used as `rmdir` - https://github.com/planetscale/vitess/blob/tal_vtgate_cluster/go/test/endtoend/cluster/etcd_process.go#L130. That is why I need to pass the `command`.

deepthi replied:
Then there is a bug in ManageTopoDir because that is always calling etcdctl with mkdir.

arindamnayak replied:
This will be fixed in next commit.

deepthi commented on go/test/endtoend/cluster/etcd_process.go:
```diff
@@ -0,0 +1,160 @@
+
+// ManageTopoDir creates global and zone in etcd2
+func (etcd *EtcdProcess) ManageTopoDir(command string, directory string) error {
+	tmpProcess := exec.Command(
```
you can save time by naming such variables with single letters - like `p` for process. That is acceptable.

deepthi commented on go/test/endtoend/cluster/vtctlclient_process.go:
```diff
@@ -0,0 +1,105 @@
+		vtctlclient.Binary,
+		"-server", vtctlclient.Server,
+		"ApplySchema",
+		"-sql-file", file.Name(),
```
I don't think you need to create a file and pass it. `ApplySchema` has an option `-sql` where you can directly provide the schema sql string.

arindamnayak replied:
Changed code to use `SQL` directly instead of file.

deepthi commented on go/test/endtoend/cluster/vtctlclient_process.go:
```diff
@@ -0,0 +1,105 @@
+	tmpProcess := exec.Command(
+		vtctlclient.Binary,
+		"-server", vtctlclient.Server,
+		"ApplyVSchema",
```
similar option here: `-vschema`

arindamnayak replied:
Changed code to use `JSON` directly instead of file.

deepthi commented on go/test/endtoend/cluster/vttablet_process.go:
```diff
@@ -0,0 +1,185 @@
+	"vitess.io/vitess/go/vt/log"
+)
+
+// VttabletProcess is a generic handle for a running vtctld .
```
`a running vttablet`

deepthi commented on go/test/endtoend/cluster/vttablet_process.go:
```diff
@@ -0,0 +1,185 @@
+	return fmt.Errorf("process '%s' timed out after 60s (err: %s)", vttablet.Name, <-vttablet.exit)
+}
+
+// WaitForStatus function checks if vtctld process is up and running
```
vtctld -> vttablet

deepthi commented on go/test/endtoend/cluster/vttablet_process.go:
```diff
@@ -0,0 +1,185 @@
+	return false
+}
+
+// TearDown shutdowns the running vttablet service
```
shutdowns -> shuts down

deepthi commented on go/test/endtoend/cluster/vttablet_process.go:
```diff
@@ -0,0 +1,185 @@
+	}
+}
+
+// VttabletProcessInstance returns a VtctlProcess handle for vtctl process
```
Fix this comment.

deepthi commented on go/test/endtoend/cluster/vtgate_process.go:
```diff
@@ -0,0 +1,171 @@
+	return false
+}
+
+// TearDown shutdowns the running vttablet service
```
shutdowns -> shuts down
vttablet -> vtgate

deepthi commented on go/test/endtoend/clustertest/main_test.go:
```diff
@@ -0,0 +1,107 @@
+)
+
+var (
+	ClusterInstance *cluster.LocalProcessCluster
```
Are any of these variables used outside this package? Or expected to be used that way?
If not, they should start with lowercase.

arindamnayak replied:
Yes, there are being using in individual testcases e.g. `vtgate_test.go` located in the same package.

deepthi replied:
If it is all in the same package, please change them to start with lowercase, which means they can't be used outside the package.

deepthi commented on go/test/endtoend/cluster/cluster_process.go:
```diff
@@ -0,0 +1,265 @@
+// StartTopo starts topology server
+func (cluster *LocalProcessCluster) StartTopo() (err error) {
+	if cluster.Cell == "" {
+		cluster.Cell = "zone1"
```
Use a constant instead of hardcoding.

deepthi reviewed (changes requested):
Nice work! I have requested a number of small changes.
Two things to note:
* You will need to add the ability to execute commands using `vtctl` or `vtctlclient` before you can write the first _real_ test case.
* Right now all log goes to stdout / stderr. That's fine, but pretty soon we will want to send the log for each process to a different file.

deepthi commented on go/test/endtoend/cluster/mysqlctl_process.go:
```diff
@@ -0,0 +1,86 @@
+	Name         string
+	Binary       string
+	LogDirectory string
+	TabletUID    int
```
Where will the actual mysql data and log be stored?

arindamnayak replied:
Mysql data is stored at : `$VTDATAROOT/vt_${tablet_uid}/data`. And mysql log is stored at `$VTDATAROOT/tmp`

deepthi replied:
What happens when tests are run in parallel? Is tablet_uid unique? otherwise, you will have multiple tests using the same data dir.

arindamnayak replied:
Yes, tablet_uid is unique. The data directory is different for each of tablets.

arindamnayak committed:
implemented pr comments

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak commented:
> Nice work! I have requested a number of small changes.
> Two things to note:
> 
> * You will need to add the ability to execute commands using `vtctl` or `vtctlclient` before you can write the first _real_ test case.
> * Right now all log goes to stdout / stderr. That's fine, but pretty soon we will want to send the log for each process to a different file.

Thanks @deepthi , I have implemented all changes you requested, also answered to the questions. I have added method to execute any custom command for `vtctlclient`. I have also added log statements, so that all standard log command line params will work e.g. `go test -log_dir=/opt/cluster_test  -stderrthreshold=INFO -alsologtostderr=true` .

arindamnayak committed:
updated vtctlclient to use execute method

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
Addd testcase for dynamic keyspace addition

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

ajeetj committed:
Seq tests in GO using new cluster

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

ajeetj committed:
added missing no auto-inc test

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

arindamnayak committed:
added log statements

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak committed:
converted existing vtgate test to use actual cluster

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak commented:
I have also ported existing vtgate test to use current cluster setup. But vstream_test.go is not working, as, I am not able to register grpc dialer. When it runs from vtgate package, that dialer was registered automatically.

deepthi commented on go/test/endtoend/cluster/vtctlclient_process.go:
```diff
@@ -73,25 +64,27 @@ func (vtctlclient *VtctlClientProcess) ApplySchema(Keyspace string, SQL string)
+		vtctlclient.Binary,
+		args...,
+	)
+	print(fmt.Sprintf("ApplyVSchema with arguments %v", strings.Join(tmpProcess.Args, " ")))
```
Change this log, it is not always ApplyVSchema.

arindamnayak replied:
I already changed this in some earlier commit.

deepthi commented:
@arindamnayak I have a couple of responses / new comments.

arindamnayak committed:
implemented pr comments #2

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

arindamnayak commented:
> @arindamnayak I have a couple of responses / new comments.

@deepthi , I have updated code, please have a look.

ajeetj committed:
review comments implemented

Signed-off-by: Ajeet jain <ajeet@planetscale.com>

ajeetj committed:
Merge pull request #16 from planetscale/tal_vtgate_seq2

VtGateV3 SEQUENCE tests in GO using new cluster

ajeetj committed:
updated sequence test package

Signed-off-by: Arindam Nayak <arindam.nayak@outlook.com>

deepthi commented:
Do not merge into planetscale/vitess:master

ajeetj commented:
Closing as we have created PR against vitess.io
